### DISCUSSION TITLE
Formatting slow in the new version?

### DISCUSSION THREAD

**Discussion by ChenCuiPlasma** [microsoft/vscode-cpptools#9087] [Q&A]
Hi,

Sorry for disturbing and this might be a naive question. I recently updated my extension from 1.7.1 to 1.9.7 and found the speed of formatting becomes obviously slower. Previously for my code when I hit Ctrl+shift+I the file was formatted almost immediately. However after my upgrade, I found the latency time for formatting obviously increased. After I hit the Ctrl+shift+I it let me wait for several seconds and a blue progress bar showed for the same file. I downgrade the extension back to 1.7.1 and this phenomena disappeared. I was wondering the reason for such phenomena and just want to inquire if there is anything that can be changed in settings to increase the speed for formatting. Thanks!

**Comment by sean-mcmanus** [comment]
We have not noticed or heard of this before. Which formatting engine are you using -- clang-format or vcFormat?

**Reply by sean-mcmanus** [reply]
Also, how many lines is the file you're using? And what OS are you using?

**Reply by sean-mcmanus** [reply]
Also, do you see a flame icon in the status bar when the formatting is happening? That would indicate that formatting is being blocked by an IntelliSense update operation.

**Reply by ChenCuiPlasma** [reply]
I tested the formatting on various files. Basically, for the files with several hundreds of lines, formatting works with no difference from previous version. However, for the files with about 3 thousand lines, there is an obvious latency for the new version.

I believe for the 1.9.7 I should use the clang-format but not sure for the 1.7.1 (most probably vcFormat). Can I know where in the settings can I switch between these two formatters?

Yes I do see a flame icon.

**Reply by sean-mcmanus** [reply]
I only see the blue bar when formatting is blocked by an IntelliSense operation. It's possible with 1.7.1 IntelliSense wasn't working correctly and that may have been fixed, or something else is blocking the main thread.

**Reply by sean-mcmanus** [reply]
Is formatting fast if you open a file, wait for the flame icon to disappear, and then format the file? If so, the slowness is caused by being blocked by an IntelliSense update operation.

**Reply by sean-mcmanus** [reply]
You can change formatting engines with the C_Cpp.formatting setting, but it's unlikely to make any difference in performance.

**Reply by ChenCuiPlasma** [reply]
Thanks. I think after changing back to vcFormat the formatting speed is not affected by the flame icon while when using the clangFormat the speed is obviously affected by the flame icon. Even after the flame icon disapperas the clangFormat still return me a blue bar and let me wait for a while.

**Reply by ChenCuiPlasma** [reply]
Also it seems the format return by the vcFormat is slightly different than the format return by the clangFormat. Is there anyway to make them the same to make the code consistent? Thanks

**Reply by sean-mcmanus** [reply]
vcFormat settings are controlled by the C_Cpp.vcFormat settings. The formatting engines are different so it's not possible to make them match 100%.

You may be seeing clangFormat be slower because it launches a 2nd process, which might be delayed in some cases, such as if a tag parsing operation is occurring. That delay might be fixed in our next release.

**Reply by ChenCuiPlasma** [reply]
great, thank you. I will use the 1.7.1 version at this time and wait for the next release.

**Reply by sean-mcmanus** [reply]
Is there a reason you're using 1.7.1 and not 1.8.4?

**Reply by ChenCuiPlasma** [reply]
I think the 1.8.4 shows the same problem as 1.9.7 on the latency of the formatting. 1.7.1 is the most updated version that I found no latency on the formatting. Thanks.

**Reply by sean-mcmanus** [reply]
The delay I was thinking you may be seeing in 1.9.7 was an added lock to fix https://github.com/microsoft/vscode-cpptools/issues/8764 , however, that lock should only be held for a short period of time during which we initially launch the cpptools-srv process, so I don't think it would be causing the delay you're seeing, so the alternative process creation mechanism in our next release may not fix the issue you're seeing. You could potentially attach a debugger to the cpptools process while it's stuck that could help determine the exact call stack that is causing the delay: https://github.com/microsoft/vscode-cpptools/wiki/Attaching-debugger-to-cpptools-or-cpptools%E2%80%90srv . Also, 1.8.4 doesn't have that lock, so if you repro the issue in 1.8.4 then it'd be some issue starting with that release.

**Reply by sean-mcmanus** [reply]
Okay, then we're not aware of any change in 1.8.4 that could cause a regression in clang-format performance, and we haven't seen other user reports of this and/or seen a repro ourselves. We might need more repro details in order to fix it.

**Reply by ChenCuiPlasma** [reply]
Can I know how to output the log info based on the method in the link you provided? I was able to attach the gdb to the progress but I am not sure where to output the logs.

**Reply by sean-mcmanus** [reply]
What OS are you using? After attaching, you would use the Pause button in VS Code to break into the debugger and then you can copy the call stack for the thread that has "message_handler::handle_message" on the call stack, which is the thread that is supposed to handle the format request.

**Reply by ChenCuiPlasma** [reply]
Thanks, I will try it this weekend. I am using the Ubuntu 16.04.

**Reply by sean-mcmanus** [reply]
The expected call stack on Linux would look like 
![image](https://user-images.githubusercontent.com/19859882/160205957-7b0a283a-24b9-4181-9d5c-820c5b17fbf2.png)

If you see cpptools_format on the call stack then we're stuck waiting for clang-format to finish -- otherwise it would indicate the thread is blocked by some other operation.

If it finishes too fast to break into the debugger, you could try copy/pasting the file multiple times; however, that could cause it to be break into the debugger at a different point than normal, so you might want to compare the repro and/or call stack with 1.7.1.

**Reply by ChenCuiPlasma** [reply]
thanks, I will try this out.
